Sync/Synchronizers/Bookmarks/BookmarksSynchronizer.swift

```diff
@@ -118,26 +118,34 @@ open class BufferingBookmarksSynchronizer: TimestampedSingleCollectionSynchroniz
         }
     }
 
-    func buildMobileRootAndChildrenRecords(_ local: LocalItemSource, _ buffer: BufferItemSource, additionalChildren: [BookmarkMirrorItem]) -> Deferred<Maybe<(mobileRootRecord: Record<BookmarkBasePayload>, childrenRecords: [Record<BookmarkBasePayload>])>> {
-        let childrenRecords = additionalChildren.map { bkm -> Record<BookmarkBasePayload> in
+    func buildMobileRootAndChildrenRecords(_ local: LocalItemSource, _ buffer: BufferItemSource, additionalChildren: [BookmarkMirrorItem], deletedChildren: [GUID]) -> Deferred<Maybe<(mobileRootRecord: Record<BookmarkBasePayload>, childrenRecords: [Record<BookmarkBasePayload>])>> {
+        let childrenRecords =
+        additionalChildren.map { bkm -> Record<BookmarkBasePayload> in
             let payload = bkm.asPayload()
             let mappedGUID = payload["id"].string ?? bkm.guid
             return Record<BookmarkBasePayload>(id: mappedGUID, payload: payload)
+        } +
+        deletedChildren.map { guid -> Record<BookmarkBasePayload> in
+            let payload = BookmarkBasePayload.deletedPayload(guid)
+            let mappedGUID = payload["id"].string ?? guid
+            return Record<BookmarkBasePayload>(id: mappedGUID, payload: payload)
         }
 
-        return self.buildMobileRootRecord(local, buffer, additionalChildren: additionalChildren) >>== { mobileRootRecord in
+        return self.buildMobileRootRecord(local, buffer, additionalChildren: additionalChildren, deletedChildren: deletedChildren) >>== { mobileRootRecord in
             return deferMaybe((mobileRootRecord: mobileRootRecord, childrenRecords: childrenRecords))
         }
     }
 
     func uploadSomeLocalRecords(_ storage: SyncableBookmarks & LocalItemSource & MirrorItemSource, _ mirrorer: BookmarksMirrorer, _ bookmarksClient: Sync15CollectionClient<BookmarkBasePayload>, mobileRootRecord: Record<BookmarkBasePayload>, childrenRecords: [Record<BookmarkBasePayload>]) -> Success {
         let records = [mobileRootRecord] + childrenRecords
-        let newBookmarkGUIDs = childrenRecords.map { $0.id }
+        let newBookmarkGUIDs = childrenRecords.filter { !$0.payload.deleted }.map { $0.id }
+        let deletedBookmarksGUIDs = childrenRecords.filter { $0.payload.deleted }.map { $0.id }
```

Comment:
Can we avoid doing this filter twice?

Perhaps pass the deleted and non-deleted children separately. After all, they're very different, and come from different places.